File: SHIPBOB_TO_GORGIAS/go/cron.go
package main

import (
	"context"
	"sort"
	"strings"
	"time"
)

var cronLogger = newLogger("cron")

// The cron flow: ShipBob orders -> Gorgias customer sidebar.
//
// Keeps each Gorgias customer's data.pandium.shipbob_orders in sync with that
// customer's recent ShipBob orders. Runs on a schedule and resumes where the last
// run left off, using tenant metadata as the cursor.
//
// The run is bounded at ~10 minutes by Pandium. To stay resumable, the loop keeps
// cursor state current as each order is processed, and a self-imposed deadline
// (ctx, below) stops the loop before the hard kill. Stopping there and returning
// the cursor normally is still success: the partial cursor is merged into
// metadata and the next run picks up from there.
//
// The two cursors resume differently. new_order_start_date climbs per order over
// an oldest-first query, so it is sound wherever the run stops.
// updated_order_start_date is the minimum across every page, so it only holds
// once the query is exhausted — an unread page can carry an older update — and a
// run cut short leaves it where it started. Re-syncing what it covers again is
// harmless: customer writes are idempotent PUTs.
const (
	alarmDuration   = 540 * time.Second // self-imposed 9-min deadline, ahead of Pandium's ~10-min kill
	oneMonth        = 30 * 24 * time.Hour
	maxOrdersToSync = 10 // most recent N orders kept on each customer
)

// clamp keeps a cursor within [now - 1 month, now]. Unparseable/missing values
// fall back to one month ago (the oldest window we ever fetch).
func clamp(value string, now time.Time) time.Time {
	floor := now.Add(-oneMonth)
	parsed, ok := parseTimestamp(value)
	if !ok {
		return floor
	}
	if parsed.Before(floor) {
		return floor
	}
	if parsed.After(now) {
		return now
	}
	return parsed
}

func formatCursor(t time.Time) string {
	return t.UTC().Format("2006-01-02T15:04:05.000000")
}

// cursorState is the record written back as this run's cursor.
type cursorState struct {
	newOrderStartDate     string
	updatedOrderStartDate string
}

func (c *cursorState) snapshot() map[string]any {
	return map[string]any{
		"new_order_start_date":     c.newOrderStartDate,
		"updated_order_start_date": c.updatedOrderStartDate,
	}
}

func orderID(order map[string]any) float64 {
	return toFloat64(order["id"])
}

func toFloat64(v any) float64 {
	switch n := v.(type) {
	case float64:
		return n
	case int:
		return float64(n)
	}
	return 0
}

func trimTo(s string, n int) string {
	if len(s) > n {
		return s[:n]
	}
	return s
}

// upsertOrder merges orderPayload into a customer's order list (replace by id,
// else append), then sorts and trims to the most recent maxOrdersToSync.
func upsertOrder(orders []map[string]any, orderPayload map[string]any, newestFirst bool) []map[string]any {
	newID := orderID(orderPayload)
	for i, existing := range orders {
		if orderID(existing) == newID {
			orders[i] = orderPayload
			return orders // in-place replace; no re-sort/trim needed
		}
	}

	orders = append(orders, orderPayload)
	sort.Slice(orders, func(i, j int) bool {
		if newestFirst {
			return orderID(orders[i]) > orderID(orders[j])
		}
		return orderID(orders[i]) < orderID(orders[j])
	})
	if len(orders) > maxOrdersToSync {
		if newestFirst {
			orders = orders[:maxOrdersToSync]
		} else {
			orders = orders[len(orders)-maxOrdersToSync:]
		}
	}
	return orders
}

// processOrder finds-or-creates the order's Gorgias customer, then PUT/POSTs its
// updated data.pandium.shipbob_orders. cache accumulates customer payloads within
// a run so multiple orders for one customer batch onto the same record.
func processOrder(ctx context.Context, order map[string]any, gorgias GorgiasClient, cache map[string]map[string]any, newestFirst bool) {
	r := recipientFromOrder(order)
	key := customerKey(r)
	email := validEmail(r.Email)

	customer, cached := cache[key]
	if !cached {
		var externalID string
		if email == "" {
			externalID = key
		}
		existing, err := gorgias.FindCustomer(ctx, email, externalID)
		if err != nil {
			cronLogger.Error("cannot fetch customer; skipping order", "order_id", formatID(order["id"]), "customer_key", key, "error", err)
			return
		}

		if existing != nil {
			// Anything already under data.pandium came from outside this
			// integration — a hand-edited customer can carry {"pandium": null} —
			// so check the type at every level rather than just the leaf.
			data, ok := existing["data"].(map[string]any)
			if !ok {
				data = map[string]any{}
			}
			pandium, ok := data["pandium"].(map[string]any)
			if !ok {
				pandium = map[string]any{}
			}
			if _, ok := pandium["shipbob_orders"].([]any); !ok {
				pandium["shipbob_orders"] = []any{}
			}
			data["pandium"] = pandium
			customer = map[string]any{"id": existing["id"], "data": data}
		} else {
			customer = newCustomerPayload(r, key)
		}
		cache[key] = customer
	}

	data, _ := customer["data"].(map[string]any)
	pandium, _ := data["pandium"].(map[string]any)
	ordersAny, _ := pandium["shipbob_orders"].([]any)
	orders := make([]map[string]any, 0, len(ordersAny))
	for _, o := range ordersAny {
		if m, ok := o.(map[string]any); ok {
			orders = append(orders, m)
		}
	}
	orders = upsertOrder(orders, orderDataPayload(order), newestFirst)
	ordersAny = make([]any, len(orders))
	for i, o := range orders {
		ordersAny[i] = o
	}
	pandium["shipbob_orders"] = ordersAny

	if id, hasID := customer["id"]; hasID {
		if err := gorgias.UpdateCustomer(ctx, toFloat64(id), customer); err != nil {
			cronLogger.Error("failed to upsert Gorgias customer", "customer_key", key, "error", err)
		}
	} else {
		newID, err := gorgias.CreateCustomer(ctx, customer)
		if err != nil {
			cronLogger.Error("failed to upsert Gorgias customer", "customer_key", key, "error", err)
			return
		}
		customer["id"] = newID
	}
}

type cronDeps struct {
	ShipBob ShipBobClient
	Gorgias GorgiasClient
	Now     time.Time
}

func cronRun(pandium *Pandium) (map[string]any, error) {
	shipbob, err := NewShipBobAPI(pandium)
	if err != nil {
		return nil, err
	}
	gorgias, err := NewGorgiasAPI(pandium)
	if err != nil {
		return nil, err
	}
	ctx, cancel := context.WithTimeout(context.Background(), alarmDuration)
	defer cancel()
	return runCron(ctx, pandium, cronDeps{
		ShipBob: shipbob,
		Gorgias: gorgias,
		Now:     time.Now(),
	})
}

// runCron is the tested core: everything it touches arrives through ctx/deps. A
// run that stops itself early because ctx's deadline passed is still a success —
// it returns the cursor normally, the same as a run that finished on its own.
func runCron(ctx context.Context, pandium *Pandium, deps cronDeps) (map[string]any, error) {
	now := deps.Now
	metadata := pandium.Metadata()
	if metadata == nil {
		metadata = map[string]any{}
	}
	fallback := pandium.Config["order_start_date"]

	newCursor := clamp(firstNonEmpty(asString(metadata["new_order_start_date"]), fallback), now)
	updatedCursor := clamp(firstNonEmpty(asString(metadata["updated_order_start_date"]), fallback), now)

	state := &cursorState{
		newOrderStartDate:     formatCursor(newCursor),
		updatedOrderStartDate: formatCursor(updatedCursor),
	}

	cache := make(map[string]map[string]any)
	newestFirst := strings.ToLower(pandium.Config["newest_order_first"]) == "true"

	// New orders: SortOrder=Oldest, so created_date advances forward monotonically.
	// The cursor is written per order (below), so stopping anywhere in this loop
	// leaves it at a sound value — no separate "did we finish" bookkeeping needed.
	cronLogger.Info("syncing new ShipBob orders", "start_date", state.newOrderStartDate)
	page := 1
newOrdersLoop:
	for {
		if ctx.Err() != nil {
			break
		}
		orders, err := deps.ShipBob.NewOrdersPage(ctx, newCursor, page)
		if err != nil {
			if ctx.Err() != nil {
				break // the deadline firing mid-request looks like a fetch error; treat it as a clean stop
			}
			return nil, err
		}
		if len(orders) == 0 {
			break
		}
		for _, order := range orders {
			if ctx.Err() != nil {
				break newOrdersLoop
			}
			cronLogger.Info("processing new order", "order_id", formatID(order["id"]))
			processOrder(ctx, order, deps.Gorgias, cache, newestFirst)
			// created_date is YYYY-MM-DDThh:mm:ss.sssssss+00:00; trim to 26 chars
			// for a valid (naive, microsecond) date-time.
			if created, ok := order["created_date"].(string); ok && created != "" {
				state.newOrderStartDate = trimTo(created, 26)
			}
		}
		page++
	}

	// Updated orders: keyed off shipment last_update_at (see UpdateDate).
	cronLogger.Info("syncing updated ShipBob orders", "start_date", state.updatedOrderStartDate)
	page = 1
	// Each page is sorted newest-first, but pages are not sorted relative to each
	// other, so the cursor is the minimum across every processed order — not
	// whatever the last order of the last page happened to carry. Kept in a local
	// variable, not cursorState, until the loop ends: every update date is, by
	// construction, later than the starting cursor, so folding that in would pin
	// the cursor there forever, and a partial minimum would sit newer than the
	// pages still unread. exhausted tracks whether that end was actually reached —
	// stopping early (the deadline passing mid-loop) must leave the cursor where it started.
	var oldestUpdate *time.Time
	exhausted := false
updatedOrdersLoop:
	for {
		if ctx.Err() != nil {
			break
		}
		orders, err := deps.ShipBob.UpdatedOrdersPage(ctx, updatedCursor, page)
		if err != nil {
			if ctx.Err() != nil {
				break
			}
			return nil, err
		}
		if len(orders) == 0 {
			exhausted = true
			break
		}
		for _, order := range orders {
			if ctx.Err() != nil {
				break updatedOrdersLoop
			}
			cronLogger.Info("processing updated order", "order_id", formatID(order["id"]))
			processOrder(ctx, order, deps.Gorgias, cache, newestFirst)
			updateDate := deps.ShipBob.UpdateDate(order, updatedCursor)
			if oldestUpdate == nil || updateDate.Before(*oldestUpdate) {
				oldestUpdate = &updateDate
			}
		}
		page++
	}

	// Every page is in, so the minimum is final and safe to resume from.
	if exhausted && oldestUpdate != nil {
		state.updatedOrderStartDate = formatCursor(*oldestUpdate)
	}

	return state.snapshot(), nil
}

func firstNonEmpty(values ...string) string {
	for _, v := range values {
		if v != "" {
			return v
		}
	}
	return ""
}

File: shipbob_to_gorgias/rust/src/cron.rs
//! The cron flow: ShipBob orders → the Gorgias customer sidebar.
//!
//! Keeps each Gorgias customer's `data.pandium.shipbob_orders` in sync with that
//! customer's recent ShipBob orders. Runs on a schedule and resumes where the
//! last run left off, using a cursor stored in tenant metadata.
//!
//! Pandium bounds a run at roughly ten minutes, so a tenant with a large backlog
//! will not finish in one pass. To stay resumable, the sync keeps a single
//! [`Cursors`] value current as each order is processed, and a watchdog thread
//! flushes it if the run gets close to the limit. Ending the run successfully in
//! that case is the whole point: Pandium merges a *successful* run's metadata,
//! so the next run reads the partial cursor and picks up from there. A run that
//! hits Pandium's own limit instead is marked **Failed (Timeout)** and writes
//! nothing.
//!
//! What a partial run may commit differs per half; see [`Cursors`].

use std::collections::HashMap;
use std::process;
use std::sync::{Arc, Mutex};
use std::thread;
use std::time::Duration;

use anyhow::Result;
use chrono::{NaiveDateTime, TimeDelta, Utc};
use serde_json::{Value, json};

use crate::dates;
use crate::gorgias::{CustomerKey, Gorgias, Helpdesk};
use crate::pandium::{self, Pandium};
use crate::shipbob::{Orders, Recipient, ShipBob};

/// Nine minutes: a self-imposed deadline a minute inside Pandium's limit, which
/// leaves room to write the cursor before the platform stops the run.
pub const DEADLINE: Duration = Duration::from_secs(540);

/// How far back the very first sync may reach, and the floor every later cursor
/// is held to.
const MAX_LOOKBACK_DAYS: i64 = 30;

/// How many of a customer's most recent orders the sidebar keeps.
const MAX_ORDERS_TO_SYNC: usize = 10;

/// The point each of the two order queries resumes from.
///
/// The two halves are written at different times, because their queries order
/// results differently:
///
/// * `new_orders` is fetched oldest-first, so `created_date` climbs as the sync
///   goes. It advances per order and is sound to flush at any moment.
/// * `updated_orders` is the *minimum* update across the whole result set, which
///   is not known until the last page is read. It is written once the loop ends,
///   so a run cut short leaves it where the run found it.
///
/// Either way it tracks orders *attempted*, not orders Gorgias accepted:
/// `process_order` logs a write failure and moves on, so one unreachable
/// customer costs that customer's order rather than the rest of the backlog.
#[derive(Debug, Clone, Copy)]
pub struct Cursors {
    pub new_orders: NaiveDateTime,
    pub updated_orders: NaiveDateTime,
}

impl Cursors {
    /// The cursor as Pandium stores it. Only these two keys are written, so the
    /// shallow merge leaves the webhook flow's `processed_events` untouched.
    pub fn as_metadata(&self) -> Value {
        json!({
            "new_order_start_date": dates::iso(self.new_orders),
            "updated_order_start_date": dates::iso(self.updated_orders),
        })
    }
}

/// Hold a cursor inside `[now - 30 days, now]`. A missing or unparseable value —
/// a first run, mostly — starts at the floor, the oldest window ever fetched.
pub fn clamp(value: Option<&str>, now: NaiveDateTime) -> NaiveDateTime {
    let floor = now - TimeDelta::days(MAX_LOOKBACK_DAYS);
    value
        .and_then(dates::parse)
        .map_or(floor, |value| value.clamp(floor, now))
}

pub fn run(pandium: &Pandium) -> Result<Value> {
    let now = Utc::now().naive_utc();
    let metadata = pandium.metadata();
    // The end user supplies the start date from the connection settings form until the first
    // run has written a cursor of its own.
    let configured = pandium
        .config("order_start_date")
        .filter(|start| !start.is_empty());
    let stored = |key: &str| {
        metadata[key]
            .as_str()
            .filter(|start| !start.is_empty())
            .map(str::to_string)
    };

    let cursors = Arc::new(Mutex::new(Cursors {
        new_orders: clamp(
            stored("new_order_start_date").as_deref().or(configured),
            now,
        ),
        updated_orders: clamp(
            stored("updated_order_start_date").as_deref().or(configured),
            now,
        ),
    }));
    flush_at_deadline(Arc::clone(&cursors), DEADLINE);

    let shipbob = ShipBob::new(pandium)?;
    let gorgias = Gorgias::new(pandium)?;
    let newest_first = pandium.config_flag("newest_order_first");

    sync(&shipbob, &gorgias, &cursors, newest_first, now)?;

    // Reached the end in time; the watchdog dies with the process on return.
    let cursors = *cursors.lock().unwrap();
    Ok(cursors.as_metadata())
}

/// Flush the cursor and end the run successfully if the sync is still going when
/// `deadline` passes.
///
/// This is the Rust shape of the pattern: the sync loop and the watchdog share
/// one cursor behind a `Mutex`, so the watchdog always writes whatever the loop
/// had reached. The thread is never joined — it is only there for the run that
/// does not finish, and it dies with the process when the run that does returns.
fn flush_at_deadline(cursors: Arc<Mutex<Cursors>>, deadline: Duration) {
    thread::spawn(move || {
        thread::sleep(deadline);
        log::warn!("approaching the run-time limit — flushing the cursor for the next run");
        // The same writer the normal path uses, so there is exactly one route to
        // stdout, and exit 0 so Pandium counts the run as a success and merges
        // the partial cursor.
        pandium::update_metadata(&cursors.lock().unwrap().as_metadata());
        process::exit(0);
    });
}

/// Run both halves of the sync, advancing `cursors` as each order is processed.
///
/// A failed ShipBob fetch returns `Err`, ending the run without writing
/// metadata, so the next run resumes from the last cursor a completed run
/// stored. Re-syncing what it covers again is harmless: the customer write is an
/// idempotent PUT.
///
/// Split out from [`run`] so it can be driven by test doubles: everything it
/// touches arrives through the two traits.
pub fn sync(
    shipbob: &dyn Orders,
    gorgias: &dyn Helpdesk,
    cursors: &Mutex<Cursors>,
    newest_first: bool,
    now: NaiveDateTime,
) -> Result<()> {
    // Orders for each customer batch onto a single record.
    let mut customers = HashMap::new();

    // New orders come back oldest-first, so created_date advances monotonically
    // and the last order processed is the right place to resume from.
    let start = cursors.lock().unwrap().new_orders;
    log::info!("syncing new ShipBob orders since {}", dates::iso(start));
    for page in 1.. {
        let orders = shipbob.new_orders_page(start, page)?;
        if orders.is_empty() {
            break;
        }
        for order in &orders {
            log::info!("processing new order with id {}", order["id"]);
            process_order(order, gorgias, &mut customers, newest_first);
            if let Some(created) = order["created_date"].as_str().and_then(dates::parse) {
                cursors.lock().unwrap().new_orders = created;
            }
        }
    }

    // Updated orders are sorted newest-first within a page but not across pages,
    // so the cursor is the *minimum* over every order processed — not whatever
    // the last one happened to carry. Kept out of `cursors` until the loop ends:
    // every update is by construction later than the starting point, so folding
    // that in would pin the cursor there forever, and a partial minimum would
    // sit newer than the pages still unread.
    let start = cursors.lock().unwrap().updated_orders;
    log::info!("syncing updated ShipBob orders since {}", dates::iso(start));
    let mut oldest_update: Option<NaiveDateTime> = None;
    for page in 1.. {
        let orders = shipbob.updated_orders_page(start, page)?;
        if orders.is_empty() {
            break;
        }
        for order in &orders {
            log::info!("processing updated order with id {}", order["id"]);
            process_order(order, gorgias, &mut customers, newest_first);

            let updated = crate::shipbob::update_date(order, start, now);
            if oldest_update.is_none_or(|oldest| updated < oldest) {
                oldest_update = Some(updated);
            }
        }
    }

    // Every page is in, so the minimum is final and safe to resume from.
    if let Some(oldest) = oldest_update {
        cursors.lock().unwrap().updated_orders = oldest;
    }

    Ok(())
}

/// Find-or-create the order's Gorgias customer, then write their updated
/// `data.pandium.shipbob_orders` back.
///
/// Every Gorgias failure here is logged and swallowed so that the sync keeps
/// going. The order is not picked up again on the next run: see [`Cursors`] for
/// what that means for the cursor.
fn process_order(
    order: &Value,
    gorgias: &dyn Helpdesk,
    customers: &mut HashMap<String, Value>,
    newest_first: bool,
) {
    let recipient = Recipient::of(order);
    let key = CustomerKey::for_recipient(&recipient);

    if !customers.contains_key(key.as_str()) {
        let found = match gorgias.find_customer(&key) {
            Ok(found) => found,
            Err(err) => {
                log::error!(
                    "skipping order {} — cannot fetch customer {}: {err:#}",
                    order["id"],
                    key.as_str()
                );
                return;
            }
        };
        let customer = match found {
            Some(customer) => json!({"id": customer["id"], "data": sidebar_data(&customer)}),
            None => crate::gorgias::new_customer_payload(&recipient, &key),
        };
        customers.insert(key.as_str().to_string(), customer);
    }

    let customer = customers.get_mut(key.as_str()).expect("just inserted");
    upsert(
        &mut customer["data"]["pandium"]["shipbob_orders"],
        crate::gorgias::order_entry(order),
        newest_first,
    );

    let result = match customer["id"].as_i64() {
        Some(id) => gorgias.update_customer(id, customer),
        None => gorgias.create_customer(customer).map(|id| {
            // Remember the new id so the next order for this customer updates
            // the record instead of creating a second one.
            customer["id"] = id.into();
        }),
    };
    if let Err(err) = result {
        log::error!(
            "failed to upsert Gorgias customer {}: {err:#}",
            key.as_str()
        );
    }
}

/// An existing customer's `data`, with the path this integration owns made safe
/// to write to.
///
/// Anything already under `data.pandium` came from outside this integration — a
/// hand-edited customer can carry `{"pandium": null}` — so every level is
/// checked rather than just the leaf, and keys the integration does not own are
/// left where they are.
fn sidebar_data(customer: &Value) -> Value {
    let mut data = customer["data"].clone();
    if !data.is_object() {
        data = json!({});
    }
    if !data["pandium"].is_object() {
        data["pandium"] = json!({});
    }
    if !data["pandium"]["shipbob_orders"].is_array() {
        data["pandium"]["shipbob_orders"] = json!([]);
    }
    data
}

/// Merge `entry` into a customer's order list — replacing the order with the
/// same id, or adding it and trimming the list back to the most recent
/// [`MAX_ORDERS_TO_SYNC`].
fn upsert(orders: &mut Value, entry: Value, newest_first: bool) {
    let Some(orders) = orders.as_array_mut() else {
        return;
    };

    if let Some(existing) = orders.iter_mut().find(|order| order["id"] == entry["id"]) {
        *existing = entry; // replaced in place: order and length are unchanged
        return;
    }

    orders.push(entry);
    orders.sort_by_key(|order| order["id"].as_i64().unwrap_or_default());
    if newest_first {
        orders.reverse();
    }
    if orders.len() > MAX_ORDERS_TO_SYNC {
        // The list is sorted, so the orders to drop are always at the far end.
        match newest_first {
            true => orders.truncate(MAX_ORDERS_TO_SYNC),
            false => drop(orders.drain(..orders.len() - MAX_ORDERS_TO_SYNC)),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::fakes::{self, RecordingGorgias};

    fn at(value: &str) -> NaiveDateTime {
        dates::parse(value).expect("a test timestamp")
    }

    fn cursors_from(start: &str) -> Arc<Mutex<Cursors>> {
        Arc::new(Mutex::new(Cursors {
            new_orders: at(start),
            updated_orders: at(start),
        }))
    }

    /// Run the sync the way a tenant with the default settings would: oldest
    /// first, and a `now` comfortably later than every timestamp in the fixtures.
    fn run_sync(
        shipbob: &dyn Orders,
        gorgias: &dyn Helpdesk,
        cursors: &Mutex<Cursors>,
    ) -> Result<()> {
        sync(shipbob, gorgias, cursors, false, at("2026-07-20T00:00:00"))
    }

    #[test]
    fn the_sync_pages_until_empty_and_keeps_the_cursor_current_as_it_goes() {
        // Advancing the cursor per order rather than once at the end is what
        // makes the flow resumable.
        let shipbob = fakes::ShipBob::with_new_orders(vec![
            vec![fakes::order(1, "2026-07-05T10:00:00Z", Some("j@x.com"))],
            vec![fakes::order(2, "2026-07-06T10:00:00Z", Some("j@x.com"))],
        ]);
        let gorgias = RecordingGorgias::new(&[]);
        let cursors = cursors_from("2026-07-01");
        shipbob.watch(Arc::clone(&cursors));

        run_sync(&shipbob, &gorgias, &cursors).expect("every page answers");

        assert_eq!(shipbob.new_order_pages_requested(), [1, 2, 3]); // until empty

        // What a flush would have written when page 2 was fetched: order 1 done.
        assert_eq!(
            shipbob
                .cursor_when_page_fetched(2)
                .map(dates::iso)
                .as_deref(),
            Some("2026-07-05T10:00:00")
        );
        assert_eq!(
            cursors.lock().unwrap().as_metadata()["new_order_start_date"],
            "2026-07-06T10:00:00"
        );

        // Both orders batch onto one customer: created once, then updated.
        assert_eq!(gorgias.created().len(), 1);
        let (_, customer) = gorgias.updated().pop().expect("the second order updates");
        let synced = &customer["data"]["pandium"]["shipbob_orders"];
        assert_eq!(synced.as_array().unwrap().len(), 2);
    }

    #[test]
    fn the_updated_cursor_lands_on_the_oldest_update_across_every_page() {
        // Pages are each sorted newest-first, but not relative to each other, so
        // the cursor has to be the oldest update seen anywhere.
        let updated = |id, day| {
            fakes::order_updated_on(id, &format!("2026-07-{day:02}T00:00:00Z"), "j@x.com")
        };
        let shipbob = fakes::ShipBob::with_updated_orders(vec![
            vec![updated(1, 18), updated(2, 17)],
            vec![updated(3, 11), updated(4, 12)], // the oldest update overall
            vec![updated(5, 16)],                 // newer again, after the oldest page
        ]);
        let cursors = cursors_from("2026-07-01");
        shipbob.watch(Arc::clone(&cursors));

        run_sync(&shipbob, &RecordingGorgias::new(&[]), &cursors).expect("every page answers");

        assert_eq!(
            cursors.lock().unwrap().as_metadata()["updated_order_start_date"],
            "2026-07-11T00:00:00" // not order 5, the last one processed
        );

        // Until the last page is in the minimum is provisional, so a flush
        // partway through writes the value the run started with.
        for page in 1..=3 {
            assert_eq!(
                shipbob.updated_cursor_when_page_fetched(page).map(dates::iso),
                Some("2026-07-01T00:00:00".to_string()),
                "the cursor moved while page {page} was still outstanding"
            );
        }
    }

    #[test]
    fn a_page_that_fails_to_fetch_ends_the_run_rather_than_committing_a_cursor() {
        // A page that errors has to stay distinguishable from the empty page
        // that ends the loop, or the run would stop early and commit a cursor
        // for pages it never read.
        let updated = |id, day| {
            fakes::order_updated_on(id, &format!("2026-07-{day:02}T00:00:00Z"), "j@x.com")
        };
        let shipbob = fakes::ShipBob::with_updated_orders(vec![
            vec![updated(1, 18)],
            vec![updated(2, 11)], // never read: the fetch fails first
        ])
        .failing_on_page(2);
        let cursors = cursors_from("2026-07-01");

        let result = run_sync(&shipbob, &RecordingGorgias::new(&[]), &cursors);

        assert!(result.is_err(), "a failed fetch is not an exhausted query");
        // Stopped at the failure rather than paging on.
        assert_eq!(shipbob.updated_order_pages_requested(), [1, 2]);
    }
}
